File: SDE_sheet/Arrays/15_majority_element_nby2/solution.py
class Solution:
    def majorityElement(self, nums: List[int]) -> int:
        candidate = None
        count = 0

        for num in nums:
            if count == 0:
                candidate = num
                count = 1
            elif candidate == num:
                count += 1
            else:
                count -= 1

        count = 0
        for num in nums:
            if candidate == num:
                count += 1

        return candidate if count > len(nums) / 2 else -1

File: SDE_sheet/Arrays/15_majority_element_nby2/solution.java
class Solution {
    public int majorityElement(int[] nums) {
        Integer candidate = null;
        int count = 0;

        for (int num : nums) {
            if (count == 0) {
                candidate = num;
                count = 1;
            } else if (candidate.equals(num)) {
                count++;
            } else {
                count--;
            }
        }

        count = 0;
        for (int num : nums) {
            if (candidate.equals(num)) count++;
        }

        return count > nums.length / 2 ? candidate : -1;
    }
}
